crates/lsp/src/logging.rs

```diff
@@ -0,0 +1,324 @@
+        LogWriter::new(&self.log_tx)
+    }
+}
+
+pub(crate) fn init_logging(
+    log_tx: LogMessageSender,
+    log_level: Option<LogLevel>,
+    dependency_log_levels: Option<String>,
+    client_info: Option<&ClientInfo>,
+) {
+    let log_level = resolve_log_level(log_level);
+    let dependency_log_levels = resolve_dependency_log_levels(dependency_log_levels);
+
+    let writer = if client_info.is_some_and(|client_info| {
+        client_info.name.starts_with("Zed") || client_info.name.starts_with("Visual Studio Code")
+    }) {
+        // These IDEs are known to support `window/logMessage` well
+        BoxMakeWriter::new(LogWriterMaker::new(log_tx))
+    } else if is_test_client(client_info) {
+        // Ensures a standard `cargo test` captures output unless `-- --nocapture` is used
+        BoxMakeWriter::new(TestWriter::default())
+    } else {
+        // Fallback for other editors / IDEs
+        BoxMakeWriter::new(std::io::stderr)
+    };
+
+    let layer = tracing_subscriber::fmt::layer()
+        // Disable ANSI escapes, those are not supported in Code
+        .with_ansi(false)
+        // Display source code file paths
+        .with_file(true)
+        // Display source code line numbers
+        .with_line_number(true)
+        // Don't display the thread ID or thread name
+        .with_thread_ids(false)
+        .with_thread_names(false)
+        // Don't display the event's target (module path).
+        // Mostly redundant with file paths.
+        .with_target(false)
+        // Display local time rather than UTC
+        .with_timer(LocalTime::rfc_3339())
+        // Display the log level
+        .with_level(true)
+        .with_writer(writer)
+        .with_filter(log_filter(log_level, dependency_log_levels));
+
+    let subscriber = tracing_subscriber::Registry::default().with(layer);
+
+    if is_test_client(client_info) {
+        // During parallel testing, `set_global_default()` gets called multiple times
+        // per process. That causes it to error, but we ignore this.
+        tracing::subscriber::set_global_default(subscriber).ok();
+    } else {
+        tracing::subscriber::set_global_default(subscriber)
+            .expect("Should be able to set the global subscriber exactly once.");
+    }
+
+    tracing::info!("Logging initialized with level: {log_level}");
+}
+
+/// We use a special `TestWriter` during tests to be compatible with `cargo test`'s
+/// typical output capturing behavior.
+///
+/// Important notes:
+/// - `cargo test` swallows all logs unless you use `-- --nocapture`.
+/// - Tests run in parallel, so logs can be interleaved unless you run `--test-threads 1`.
+///
+/// We use `cargo test -- --nocapture --test-threads 1` on CI because of all of this.
+fn is_test_client(client_info: Option<&ClientInfo>) -> bool {
+    client_info.map_or(false, |client_info| client_info.name == "AirTestClient")
+}
+
+// TODO: Is there a way to generate this at compile time?
```

Comment:
chatgpt suggests

```sh
cargo metadata --no-deps --format-version 1 | jq -r '.packages[].name'
```

which gives me

```
air
air_r_formatter
air_r_syntax
air_formatter_test
line_ending
air_r_parser
air_r_factory
tests_macros
fs
lsp
lsp_test
biome_ungrammar
xtask_codegen
xtask
```

I guess we can add an exclude list for build-time things like biome_ungrammar. Or there's probably a way to determine dev deps from the metadata.

Reply:
If you remove the jq you get a bunch of json data which is fine to ingest from Rust. This could be done in `build.rs`.

Reply:
Of course there's a crate for this. `cargo_metadata`! https://github.com/posit-dev/air/pull/93/commits/3f0f72058866a26f6316dd2ad26939fea769ee8e